### PR TITLE
Add options to display type hints in docstring or signature

### LINKED ISSUE #71: Display the type hints and default values in the description.
Displaying them in the signature is too heavy. It's better fitted in the description. Same for the return type.
See https://github.com/keras-team/autokeras/pull/957#discussion_r379223556

### PR DESCRIPTION
**Pull request by leonoverweel**
This PR adds two options to `DocumentationGenerator`:

- `types_in_signature`: keeps the current behavior if `True` (default); if `False`, strips type annotations out of function signatures.
- `types_in_docstring`: keeps the current behavior if `False` (default): if `True`, inserts type annotations into argument descriptions in the docstring.

I kept these as separate options for flexibility (e.g. you may have hardcoded type hints in your docstrings already and want to strip signature types without adding the duplicate extra hints generated from the type annotations.)

Take the following method, for example:

```python
def water_plant(self, amount: Union[int, float]):
    """Give your plant some water.

    # Arguments
        amount: How much water to give
    """
    pass
```

*Current output (`types_in_signature=True, types_in_docstring=False`):*

### water_plant function

```python
tests.test_autogen.water_plant(self, amount: Union[int, float])
```
Give your plant some water.

__Arguments__

- __amount__: How much water to give

----

*Newly enabled output (`types_in_signature=False, types_in_docstring=True`):*

### water_plant function

```python
tests.test_autogen.water_plant(self, amount)
```

Give your plant some water.

__Arguments__

- __amount__: typing.Union[int, float], How much water to give

----

Closes part of #71.

### CODE REVIEW

**Comment by gabrieldemarmiesse** [comment]
Many thanks! I very quickly went over the pull request, and here are two things I'd like to change: 

* types_in_docstring should be the default, and we shouldn't be able to change it. We make that choice because it's what people will want in most cases and it's less maintenance for us. 
* Could we have the types formatted like 

* **my_parameter** `Union[soemthing, something]` : description here

**Comment by leonoverweel** [comment]
@gabrieldemarmiesse no problem, thanks for the quick review!

Just to clarify for your first point: do you want me to remove both the `types_in_signature` and `types_in_docstring` arguments, and change it to always have the new behavior?

**Comment by gabrieldemarmiesse** [comment]
Yes, you should remove both, the types should always be in the docstring (we're changing the behavior, but I wouldn't consider that a breaking change).

**Comment by leonoverweel** [comment]
Done: I removed the options and changed the format.

File: keras_autodoc/get_signatures.py [hunk 1]
```diff
@@ -1,5 +1,4 @@
 import warnings
-from sphinx.util.inspect import Signature
```
Comment (gabrieldemarmiesse) [inline keras_autodoc/get_signatures.py:2]: The advantage of using sphinx to get the signature (types hints too) compare to plain signature from `inspect` is that there are some changes to make it easier to read. I tried your pull request on the autokeras documentation, it looks great, but the types are slightly more difficult to read now:

`Optional[int]` becomes `typing.Union[int, NoneType]`

One solution may be to continue to use `Signature` from sphinx. This also makes sure we're futureproof concerning type hints. Since it's a private API of sphinx, there is no documentation, but here is the source code:  https://github.com/sphinx-doc/sphinx/blob/3.x/sphinx/util/inspect.py#L680  It might be possible to just use the function that formats the type hints. You can do what you think is best here.

File: keras_autodoc/get_signatures.py [hunk 2]
```diff
@@ -21,7 +20,15 @@ def get_signature_start(function):
...
-    signature_end = Signature(function).format_args()
+    signature = inspect.signature(function)
+    signature = signature.replace(
+        parameters=[
+            param.replace(annotation=inspect.Signature.empty)
+            for param in signature.parameters.values()
+        ],
+        return_annotation=inspect.Signature.empty
+    )
+    signature_end = str(signature)
```
Comment (gabrieldemarmiesse) [inline keras_autodoc/get_signatures.py:31]: Maybe instead we can use this: https://github.com/sphinx-doc/sphinx/blob/3.x/sphinx/util/inspect.py#L617 with `show_annotations=False`?

File: keras_autodoc/get_signatures.py [hunk 3]
```diff
@@ -21,7 +20,15 @@ def get_signature_start(function):
...
     if utils.ismethod(function):
         signature_end = signature_end.replace('(self, ', '(')
         signature_end = signature_end.replace('(self)', '()')
```
Comment (gabrieldemarmiesse) [inline keras_autodoc/get_signatures.py:34]: Just below this line, there is the substitution for aliases (for example, a type hint will display `keras.engine.base.Layer` but this will replace it with `keras.layers.Layer`). We should move it now to the place were the docstring is generated.

File: tests/test_get_signature.py [hunk 4]
```diff
@@ -14,7 +16,10 @@ def some_function(*args, **kwargs):
 
 
 class Dog:
-    def woof(self, volume, good_boy=True, name='doggy'):
+    def woof(self, volume, good_boy: bool = True, name: str = 'doggy') -> int:
+        pass
+
+    def hard_method(self, arg: Union[int, Tuple[int, int]], arg2: int = 0) -> int:
         pass
```
Comment (gabrieldemarmiesse) [inline tests/test_get_signature.py:23]: Could we have here a test with the expected type hint? `Union[int, Tuple[int, int]]` should be in the final docstring.

File: keras_autodoc/autogen.py [hunk 5]
```diff
@@ -67,9 +68,18 @@ def generate(self, dest_dir):
...
         """Can be overridden."""
-        return process_docstring(docstring)
+        processsed = process_docstring(docstring)
+
+        # Add type annotations to docstring
+        for arg, arg_type in types.items():
+            arg_type = getattr(arg_type, "__name__", arg_type)
+            processsed = re.sub(
+                rf"(- __{arg}__)", rf"\1 `{arg_type}`", processsed
+            )
```
Comment (gabrieldemarmiesse) [inline keras_autodoc/autogen.py:80]: I think this piece of code would be better fitted at this line: https://github.com/keras-team/keras-autodoc/blob/0af2f382813f047a30faadfa9dd91f7f2c070e3b/keras_autodoc/docstring.py#L80  , with a check "if the section name is `Arguments`".

File: keras_autodoc/get_signatures.py (hunk 1 above)
Comment (leonoverweel) [reply keras_autodoc/get_signatures.py:2]: Update: I think it'd more robust/future-proof to rely on Python's built-in typing support instead of using that private sphinx API. According to the Python docs, [`typing.get_type_hints()`](https://docs.python.org/3/library/typing.html#typing.get_type_hints) should enable that for us, but it doesn't seem to work for me. I'm going to [give my SO question on this](https://stackoverflow.com/questions/61320106/python-typing-get-type-hints) a few days to hopefully get a nice built-in answer, and then decide how to proceed. (I'll wait to work on your other comments until then too.)

Comment (gabrieldemarmiesse) [reply keras_autodoc/get_signatures.py:2]: If we can achieve a similar result with the public API, I'm all for it.

Comment (leonoverweel) [reply keras_autodoc/get_signatures.py:2]: Is it just the `Optional[x]` becoming `typing.Union[x, NoneType]` now that get less readable, or other things as well? It looks like [`Optional` is just syntactic sugar](https://docs.python.org/3/library/typing.html#typing.Optional) and internally gets saved as a `Union`, but if that's the only problem I think just treating that as a special case and substituting it manually is still a bit nicer than using the private sphinx API.

Comment (gabrieldemarmiesse) [reply keras_autodoc/get_signatures.py:2]: Manual substitution may not be trivial. 

There are three differences:
* Optional
* `typing.` is removed
* NoneType is displayed instead of None.

If we can acheive that with only the python public api it would be cool indeed.

Comment (leonoverweel) [reply keras_autodoc/get_signatures.py:2]: Yeah you're right; after looking at the complications in [`sphinx.util.typing.stringify()`](https://github.com/sphinx-doc/sphinx/blob/3.x/sphinx/util/typing.py#L42) I think it's probably not worth the maintenance burden to implement this ourselves. I'm now just calling that (9b73508).

File: keras_autodoc/get_signatures.py (hunk 2 above)
Comment (leonoverweel) [reply keras_autodoc/get_signatures.py:31]: Makes sense, changed in 2ab0d2f.

File: keras_autodoc/get_signatures.py (hunk 3 above)
Comment (leonoverweel) [reply keras_autodoc/get_signatures.py:27]: Class aliases seem to be used in a lot of places in that file; I'm not sure what refactor you have in mind here exactly, sorry.

File: tests/test_get_signature.py (hunk 4 above)
Comment (leonoverweel) [reply tests/test_get_signature.py:23]: Added a test in 8a38a3e.

File: keras_autodoc/autogen.py (hunk 5 above)
Comment (leonoverweel) [reply keras_autodoc/autogen.py]: Moved in 61654b8.

**Comment by leonoverweel** [comment]
Hey, I think these last few commits should cover most of the changes you requested :)

**Comment by gabrieldemarmiesse** [comment]
Many thanks! I'll take a closer look tonight or tomorrow.

File: keras_autodoc/get_signatures.py (hunk 3 above)
Comment (gabrieldemarmiesse) [reply keras_autodoc/get_signatures.py:27]: Sure thing, I'll do it after this PR is merged.

**Comment by gabrieldemarmiesse** [comment]
This is a very nice pull request, many thanks! The main thing left is to do the alias substitution, but I can do it myself in another pull request, before we do a new release, so that shouldn't be an issue.

**Comment by leonoverweel** [comment]
@gabrieldemarmiesse just wanted to follow up and mention that we've now upgraded https://docs.larq.dev to use keras-autodoc! :tada: (see larq/docs#68.)
